### PR TITLE
Fix `LOGICAL_ERROR` with functions `arrayMin/Max`

### LINKED ISSUE #69600: arrayMax / arrayMin with a function parameter not working anymore
ClickHouse 24.8.2.3

Examples
========

This taken from the documentation:

```
SELECT arrayMax(x -> (-x), [1, 2, 4]) AS res;
```

returns the error:

```
Code: 49. DB::Exception: Unexpected return type from arrayMax. Expected Int16. Got UInt8. Action: FUNCTION arrayMax(__lambda :: 1, [1, 2, 4] :: 0) -> arrayMax(lambda(tuple(x), negate(x)), [1, 2, 4]) Int16 : 3, input block structure:__lambda Function(UInt8 -> Int16) Function(size = 1), [1, 2, 4] Array(UInt8) Const(size = 1, Array(size = 1, UInt64(size = 1), UInt8(size = 3))): while executing 'FUNCTION arrayMax(__lambda :: 1, [1, 2, 4] :: 0) -> arrayMax(lambda(tuple(x), negate(x)), [1, 2, 4]) Int16 : 3'. (LOGICAL_ERROR) (version 24.8.2.3 (official build)) 
```
But the error persists event with this:
```
SELECT arrayMax(x -> toUInt16(-x), [1, 2, 4]) AS res;
```

This other that used to work too:
```
select arrayMax(x -> x.1, [(1, 'a'), (0, 'b')])
```

now fails with

```
Code: 49. DB::Exception: Unexpected return type from arrayMax. Expected UInt8. Got Tuple(UInt8, String).
```

### PR DESCRIPTION
**Pull request by Scartiloffista**
Resolves #69600

Small disclaimer: I don't really know CH codebase, or have a real C++ experience, this is my first PR, so *any* feedback or comment is more than welcome!

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed a `LOGICAL_ERROR` in functions `arrayMin` and `arrayMax` (issue #69600)

### CODE REVIEW

**Comment by Scartiloffista** [comment]
linked to
- https://github.com/ClickHouse/ClickHouse/issues/69600
- https://github.com/ClickHouse/ClickHouse/pull/65455

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 42c6d3d549de40b8f44e3007a9caadea449dc5d8 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/69640/42c6d3d549de40b8f44e3007a9caadea449dc5d8/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/69622/de85f5f251d52db319757d3c3324182edbbecde2/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by alfonsonishikawa** [comment]
Please add this test too:

`select arrayMax(x -> x.1, [(1, 'a'), (0, 'b')])`

**Comment by Scartiloffista** [comment]
hi @yariks5s and @rschu1ze, I'm sorry to bother you, but in the docs I've read that in case of a fuzzer fail I should ping a CH maintainer.

I've also tried to check (in a commit that I reverted) to fix as well the `getReturnType` that was changed in the old PR, but with no luck, do you have some insight about this test step?

**Comment by rschu1ze** [comment]
@Scartiloffista No problem. The fuzzer generates random queries and succeeds if they run normally (no error) or fail with a non-`LOGICAL_ERROR` return code. In other words, the fuzzer tries to generate queries which cause `LOGICAL_ERRORS` - this indicates something goes terribly wrong internally.

Let's go to one of the failing fuzzer tests: https://s3.amazonaws.com/clickhouse-test-reports/69640/c3274e24c0c3e36e6e514ba6e067acfa1417acd3/ast_fuzzer__tsan_.html

Click "fatal.log" and find

```sql
SELECT arrayMin(x1 -> (x1 * toNullable(-1)), materialize([1, 2, 3])) GROUP BY [materialize(3)], arrayMin([arrayMin([1, materialize(2), 3, 6]), 5, arrayMax([5])]), [2, 3, 4, 5, 6]
```

I did not test but I assume it fails with an `LOGICAL_ERROR` on this PR. Try to [minimize](https://en.wikipedia.org/wiki/Minimal_reproducible_example) it as much as possible, then check how to fix it.

**Comment by Scartiloffista** [comment]
it seems this particular query fails also with the same error on latest :( 

https://fiddle.clickhouse.com/4519542d-d086-49c8-a198-cda01a955014

also with version from before #65455, but that's normal...
https://fiddle.clickhouse.com/a266c835-d231-4cef-b5d2-dceefd3247ad

i guess this should be fixed anyway, even if it's not "dependent" on this pr

**Comment by yariks5s** [comment]
Yes, as this PR is a bit related to this topic it's okay to resolve this problem here. If you have some ideas on how to solve it, it would be wonderful🙂

**Comment by Scartiloffista** [comment]
I can try. My only worry is that the query proposed by the fuzzer in no version, before or after the PR who introduced the error, works.

Basically my idea is to continue on something like https://github.com/ClickHouse/ClickHouse/pull/69640/commits/3d4158890f59568de2d8402beda7c066680f9e8f, but even if I check for numbers and nullable who contains number, and skip that if, it would still fail. 

I guess it can be done somewhere inside the `auto call = [&](const auto & types)`, I'll try that 🤷

File: src/Functions/array/arrayAggregation.cpp [hunk 1]
```diff
@@ -360,7 +360,7 @@ struct ArrayAggregateImpl
 
     static ColumnPtr execute(const ColumnArray & array, ColumnPtr mapped)
     {
-        if constexpr (aggregate_operation == AggregateOperation::max || aggregate_operation == AggregateOperation::min)
+        if (!mapped->isNumeric() && !isNumber(array.getDataPtr()->getDataType()) && (aggregate_operation == AggregateOperation::max || aggregate_operation == AggregateOperation::min))
```
Comment (alfonsonishikawa) [inline src/Functions/array/arrayAggregation.cpp:363]: I don't know the inners of the logic, but I suggest to put the `!mapped->isNumeric()&& ...` inside the `if constexpr (...` body block.

The `if constexpr` is evaluated at compile time, but the `!mapped->isNumeric() && ...` is evaluated at runtime.

File: tests/queries/0_stateless/01602_array_aggregation.sql [hunk 2]
```diff
@@ -9,6 +9,9 @@ SELECT arrayMin([[3], [1], [2]]);
 SELECT 'Array max :';
 SELECT arrayMax([[3], [1], [2]]);
 
+SELECT 'Issue #69600';
+SELECT arrayMin(x1 -> x1 * -1, [1,2,3]);
+
```
Comment (alfonsonishikawa) [inline tests/queries/0_stateless/01602_array_aggregation.sql:14]: Add this test too:

select arrayMax(x -> x.1, [(1, 'a'), (0, 'b')])
result: 1

**Comment by Algunenano** [comment]
This is just bad, but mostly because the original code is extremely wrong. The reason why this didn't work in the first place is because it's using the original column instead of  the transformed one `array`. It also duplicated code and obsoleted some code that wasn't removed. I'll send a PR dealing with all cases

**Pull request closed without merging**